source/Ship.cpp

```diff
@@ -2729,14 +2730,47 @@ double Ship::MaxReverseVelocity() const
 
 
 
+// Take environmental hazard damage from the current system.
+void Ship::DoHazard(vector<Visual> &visuals, const Hazard *hazard, double strength)
```

Comment:
I don't think this method should be in the Ship class; it belongs in Engine alongside the collision checks. Note that the CollisionSet of ships can be used to determine the subset of ships affected by a given hazard.

Reply:
CreateSparks is a private method of Ship, so either we need to make CreateSparks public, or Ship needs a method that handles taking Hazard damage so that it can create the proper effects.

Reply:
Use CollisionSet::Circle?

Reply:
That requires a radius, meaning that would require that all hazards have a range, whereas right now the range is optional and not having it means it doesn't matter where the ship is in the system. That would also make it someone clunky to deal with the minimum and maximum range, as we'd need to get all ships within the max range then all the ships within the min range and remove the ships in the min range from the max range before figuring out which ships should actually get affected.

Reply:
Using collisionset::circle should actually offer improved performance, due to the grid lookup. To support a minimum range, you would call Circle with the max range and then exclude those ships closer than the minimum range. If that isn't performant enough, we would add a new method to CollisionSet for retrieving ships within an annular slice.

I also note that a no-range-given hazard still has a range, of the fence, so you can use that. 

Another benefit of using CollisionSet is that we could (eventually) support line-of-sight hazards that can be blocked by other ships or asteroids.

Reply:
Damage still goes beyond the fence. The fence is only used as the limit of where environmental effects are created since we aren't tiling them.

Reply:
For an infinite range hazard, iterating in-system ships directly will be fine (hard to beat having explicit knowledge about the solution space). For the others, circle will be best

Reply:
I've actually gone ahead and removed infinite range hazards. A hazard with no specified range now has a max range of 10k (the system fence). This means that the environmental effects of a hazard will always indicate exactly where damage is dealt, instead of effects ending after a certain range but damage still occurring.

Reply:
Implemented a CollisionMask::Ring method so that we don't need to run CollisionMask::Circle twice and see what's in one but not the other (which, as mentioned on Discord, actually had edge case issues).